core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -4069,6 +4431,211 @@ class KafkaApis(val requestChannel: RequestChannel,
+            erroneousTopicIdPartitions.add(tp)
+            isErroneous = true
+          } else if (batch.firstOffset < prevEndOffset) {
+            erroneous += tp -> ShareAcknowledgeResponse.partitionResponse(tp, Errors.INVALID_REQUEST)
+            erroneousTopicIdPartitions.add(tp)
+            isErroneous = true
+          } else if (batch.acknowledgeTypes == null || batch.acknowledgeTypes.isEmpty) {
+            erroneous += tp -> ShareAcknowledgeResponse.partitionResponse(tp, Errors.INVALID_REQUEST)
+            erroneousTopicIdPartitions.add(tp)
+            isErroneous = true
+          } else if (batch.acknowledgeTypes.size() > 1 && batch.lastOffset - batch.firstOffset != batch.acknowledgeTypes.size() - 1) {
+            erroneous += tp -> ShareAcknowledgeResponse.partitionResponse(tp, Errors.INVALID_REQUEST)
+            erroneousTopicIdPartitions.add(tp)
+            isErroneous = true
+          } else if (batch.acknowledgeTypes.stream().anyMatch(ackType => ackType < 0 || ackType > 3)) {
+            erroneous += tp -> ShareAcknowledgeResponse.partitionResponse(tp, Errors.INVALID_REQUEST)
+            erroneousTopicIdPartitions.add(tp)
+            isErroneous = true
+          } else {
+            prevEndOffset = batch.lastOffset
+          }
+        }
+      }
+    }
+
+    erroneousTopicIdPartitions
+  }
+
+  // the callback for processing a share fetch response, invoked before throttling.
+  def processShareFetchResponse(
+                                  responsePartitionData: mutable.Map[TopicIdPartition, ShareFetchResponseData.PartitionData],
+                                  request: RequestChannel.Request,
+                                  topicIdNames : util.Map[Uuid, String],
+                                  shareFetchContext : ShareFetchContext
+                                ): ShareFetchResponse = {
+
+    val clientId = request.header.clientId
+    val versionId = request.header.apiVersion
+    val shareFetchRequest = request.body[ShareFetchRequest]
+    val groupId = shareFetchRequest.data.groupId
+    val memberId = shareFetchRequest.data.memberId
+
+    val partitions = new util.LinkedHashMap[TopicIdPartition, ShareFetchResponseData.PartitionData]
+    val nodeEndpoints = new mutable.HashMap[Int, Node]
+    responsePartitionData.foreach { case(tp, partitionData) =>
+      partitionData.errorCode match {
+        case errCode if errCode == Errors.NOT_LEADER_OR_FOLLOWER.code | errCode == Errors.FENCED_LEADER_EPOCH.code =>
+          val leaderNode = getCurrentLeader(tp.topicPartition, request.context.listenerName)
+          leaderNode.node.foreach { node =>
+            nodeEndpoints.put(node.id, node)
+          }
+          partitionData.currentLeader
+            .setLeaderId(leaderNode.leaderId)
+            .setLeaderEpoch(leaderNode.leaderEpoch)
+        case _ =>
+      }
+
+      partitions.put(tp, partitionData)
```

Comment:
Thanks for the review. Yes we could use the same, but the definition of some methods of shareFetchContext require a util.LinkedHashMap, so we would anyways require a new variable to store the converted map as it is required as an argument to multiple methods. Talking about why do we need a util.LinkedHashMap altogether, maybe we could change those method signatures to use a scala map as well, but I think that would out of scope for this PR as it would include making changes to others code as well.